### PR TITLE
Game Component Draft

### PR DESCRIPTION
**Pull request by TrueBabyChaise**
Draft of some game Component (Bomb, Character, PowerUp) + Camera

### CODE REVIEW

File: sources/Camera/Camera.hpp [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+// Edited by Louis Auzuret on 2021-05-20.
+//
+
+#pragma once
+
+#include <string>
+#include "lib/wal/sources/Entity/Entity.hpp"
+#include "lib/wal/sources/Models/Vector3.hpp"
+
+class Camera : public WAL::Entity
```
Comment (AnonymusRaccoon) [inline sources/Camera/Camera.hpp:13]: You should not create custom entities. If you want to add functionalities to an entity, add a component to the entity. This is why this is called an ECS.

File: sources/Component/Bomb/BombComponent.cpp [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+// Edited by Benjamin Henry on 2021-05-20.
+// Edited by Louis Auzuret on 2021-05-20.
+//
+
+#include "BombComponent.hpp"
+
+BombComponent::BombComponent(WAL::Entity &entity, BombType bombType, unsigned int ownerUid)
+: WAL::Component(entity)
+{
+    this->_bombType = bombType;
```
Comment (AnonymusRaccoon) [inline sources/Component/Bomb/BombComponent.cpp:12]: Initialization should be done in the initializer, not in the body.

**Review by AnonymusRaccoon — CHANGES_REQUESTED** [review]
You don't follow the ECS architecture at all.
Entities should not be derived, to add fonctionality to an entity, add it a component.

Components should do one and only one thing. That means you should have components like HealthComponent, MovableComponent, BombHolder, CollisionComponent, ExplosableComponent and so on, not a single CharacterComponent and a BombComponent that does literaly everthing.

You should not have enum with types, types should be handled by putting mutliple components together. For example, you could have a ExplosibleComponent with a bool shouldExplose and add a TimerComponent that would set the shouldExplose to true after a timer or usse a ControllableComponent that would set the shouldExplose when the user press a button.

You should design components in par with systems, not create every component you can imagine to then add systems.

For components/systems, you could create a PR for each of them. Having smaller, feature based PR will ensure that work is well separeted and that components doesn't do to many things.

Please also check your coding style, you have some indentation issues.

File: sources/Component/Health/HealthComponent.cpp [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	{
+		if (damage >= this->_healthPoint)
+			this->die();
+		else
+			this->_healthPoint -= damage;
+	}
+
+	void HealthComponent::die(void)
+	{
+		this->setDisable(true);
```
Comment (AnonymusRaccoon) [inline sources/Component/Health/HealthComponent.cpp:41]: You probably want to disable the entity and not the health component on death.

File: sources/Component/Health/HealthComponent.cpp [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+	void HealthComponent::addHealthPoint(unsigned int healthPoint)
+	{
+		this->_healthPoint += healthPoint;
+	}
+
+	void HealthComponent::takeDmg(unsigned int damage)
+	{
+		if (damage >= this->_healthPoint)
+			this->die();
```
Comment (AnonymusRaccoon) [inline sources/Component/Health/HealthComponent.cpp:34]: You should set the HP to 0 on death.

File: sources/Component/Health/HealthComponent.cpp (hunk 3 above)
Comment (TrueBabyChaise) [reply sources/Component/Health/HealthComponent.cpp:41]: Yes, I'll fix it

File: sources/Component/Controllable/ControllableComponent.hpp [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+namespace BBM 
+{
+	class ControllableComponent : public WAL::Component
+	{
+	private:
+		float _moveX = 0;
+		float _moveZ = 0;
+		bool _jump = false;
+		bool _bomb = false;
+		bool _pause = false;
```
Comment (AnonymusRaccoon) [inline sources/Component/Controllable/ControllableComponent.hpp:20]: The pause is not something that can be controlled. A controllable component allows an entity to move, not handle every input. The pause should be handled by another component/system (I believe we will have a "GameManager" entity containing components that manage the whole game state like the pause)

File: sources/Component/Controllable/ControllableComponent.hpp [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+	public:
+
+		//! @inherit
+		WAL::Component *clone(WAL::Entity &entity) const override;
+
+		//! @brief A component can't be instantiated, it should be derived.
```
Comment (AnonymusRaccoon) [inline sources/Component/Controllable/ControllableComponent.hpp:26]: Please update the documentation accordingly, I prefer not having a documentation at all then having a bad documentation.

File: sources/Component/Controllable/ControllableComponent.hpp [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+		ControllableComponent(const ControllableComponent &) = default;
+
+		//! @brief default destructor
+		~ControllableComponent() override = default;
+
+		//! @brief A component can't be assigned
+		ControllableComponent &operator=(const ControllableComponent &) = delete;
+
+		friend class KeyboardSystem;
```
Comment (AnonymusRaccoon) [inline sources/Component/Controllable/ControllableComponent.hpp:41]: Don't put friends when you don't need them. If you can't justify your use of friend, don't use them.

File: sources/Component/Health/HealthComponent.cpp [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+	{
+		if (damage >= this->_healthPoint) {
+			this->_healthPoint = 0;
+ 		} else
+			this->_healthPoint -= damage;
+	}
+
+	void HealthComponent::die(void)
+	{
+		this->_entity.setDisable(true);
```
Comment (AnonymusRaccoon) [inline sources/Component/Health/HealthComponent.cpp:41]: As we talked this morning, the die method should be removed

File: sources/Component/Health/HealthComponent.hpp [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+			//! @brief A component can't be instantiated, it should be derived.
+			HealthComponent(const HealthComponent &) = default;
+
+			//! @brief default destructor
+			~HealthComponent() override = default;
+
+			//! @brief A component can't be assigned
+			HealthComponent &operator=(const HealthComponent &) = delete;
+
+			friend class HealthSystem;
```
Comment (AnonymusRaccoon) [inline sources/Component/Health/HealthComponent.hpp:49]: Stop using friends

File: sources/Component/Keyboard/KeyboardComponent.hpp [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+			int keyLeft;
+			//! @brief move up key
+			int keyUp;
+			//! @brief move down key
+			int keyDown;
+
+			//! @inherit
+			WAL::Component *clone(WAL::Entity &entity) const override;
+
+			//! @brief A component can't be instantiated, it should be derived.
```
Comment (AnonymusRaccoon) [inline sources/Component/Keyboard/KeyboardComponent.hpp:35]: Docs

**Comment by Octopus773** [comment]
@EternalRat  You should resolve conversations when the issue is fixed
I thought you still had issues until I looked commit by commit changes

**Review by Octopus773 — APPROVED** [review]

**Comment by AnonymusRaccoon** [comment]
This branch is stale, I based the Movement branch on commits from here so they will be on the same PR.

**Pull request closed without merging**